### ISSUE TITLE
Lumberjack server resets all tcp connections when client fails SSL negotiation

### ISSUE THREAD

**Issue by nathanlburns** [elasticsearch/logstash-forwarder#160]
This bug was originally reported in logstash, but it belongs here.

https://logstash.jira.com/browse/LOGSTASH-1598

I've scanned the issues list here (not in depth) and believe there actually may be some other problems reported here that this may explain.

Our setup:
- Many servers (Windows based) running logstash-forwarder configured to ship to round robin dns array of servers running logstash with the lumberjack input.

Our issue:
- Yesterday morning, due to a boneheaded mistake by me, the certificate I had configured lumberjack with expired, causing all logs to stop being forwarded.
- About an hour later, I had a new certificate generated and deployed on both shipper and receiver, logs started pouring in at this point.
- Expecting an initial flood of messages, I closely monitored the system, but after several hours the flood did not abate.
- Jumping into elasticsearch, realized that we were getting a ton of duplicate events, sometimes up to 100 copies.
- I added uuid to the reciever configuration and verified I was getting duplicate uuids on those messages. Problem was lumberjack related...
- logstash-forwarder logs:
```
2014/02/06 17:52:06.538400 Connected to <redacted>
2014/02/06 17:52:06.961400 Read error looking for ack: EOF
2014/02/06 17:52:06.961400 Setting trusted CA from file: <redacted>l.crt
2014/02/06 17:52:06.962400 Connecting to <redacted>
2014/02/06 17:52:07.031400 Connected to <redacted>
2014/02/06 17:52:07.033400 Socket error, will reconnect: WSASend tcp 10.0.95.65:49219: An existing connection was forcibly closed by the remote host.
```
- I ran across LOGSTASH-1598 researching these errors.
- Confirmed we still  have some windows boxes trying to use old cert.
- Identified hosts that were using bad cert using tcpdump (tcp analysis in wireshark demonstrated the rejection pattern) and blocked them with iptables.
- Duplicate messages go away.

Now I know the root problem here is a configuration issue, but existing sessions should not be closed when a client fails to connect.

I've been looking into the code a bit, currently reading the openssl implementation in ruby right now... At this point not sure how to fix it.

**Comment by maxhyjal** [comment]
Is this issue going to be fixed in logstash 1.5 release? Or is log-courier a better solution at this point at least?

**Comment by jordansissel** [comment]
I can't reproduce this on logstash master, but I can reproduce it on logstash 1.5.0 beta1. I'll try to figure out what the problem is.

**Comment by jordansissel** [comment]
strangely, I cannot reproduce it from git on v1.5.0.beta1 tag either, but I *can* reproduce it from the v1.5.0 beta1 release tarball.

**Comment by jordansissel** [comment]
On v1.5.0.beta1 I cannot reproduce the "resets all tcp connections" bug. The behavior I observe is that it closes the previously-accepted socket due to the line of code @driskell pointed out (`client` is the previous connection at exception-time due to the assignment not occurring). 

I can fix this "close the previous connection on handshake failure"

I can't yet reproduce the "closes *ALL* connections" problem this ticket mentions in the subject.

**Comment by jordansissel** [comment]
Ahh, lsf master has this already fixed, and it works, but the code is not right. I'll patch that.

**Comment by jordansissel** [comment]
I believe this is fixed and will appear in logstash 1.5.0
